### PR TITLE
Psalm Level 3 Interation.

### PR DESCRIPTION
**Pull request by alexandergull**
https://doboard.com/1/task/12367

### CODE REVIEW

File: lib/Cleantalk/ApbctWP/FindSpam/UsersChecker.php [hunk 1]
```diff
@@ -177,8 +188,8 @@ private static function removeUsersWithoutIPEmail(array $users)
             }
 
             $user_meta = self::getUserMeta($user->ID);
-
-            $user_ip    = ! empty($user_meta[0]['ip']) ? trim($user_meta[0]['ip']) : false;
+            $ip_of_user_meta = TT::getArrayValueAsString($user_meta, 'ip');
```
Comment (Glomberg) [inline lib/Cleantalk/ApbctWP/FindSpam/UsersChecker.php:191]: A bug is here: `TT::getArrayValueAsString` here will find `$user_meta['ip']` but not `$user_meta[0]['ip']` needed.

File: lib/Cleantalk/ApbctWP/FindSpam/UsersChecker.php [hunk 2]
```diff
@@ -491,17 +527,18 @@ public static function ctGetCsvFile()
             if ( is_array($user_meta) ) {
                 $user_meta = array_values($user_meta);
             }
+            $ip_of_user_meta = TT::getArrayValueAsString($user_meta, 'ip');
```
Comment (Glomberg) [inline lib/Cleantalk/ApbctWP/FindSpam/UsersChecker.php:530]: A bug is here: `TT::getArrayValueAsString` here will find `$user_meta['ip']` but not `$user_meta[0]['ip']` needed.

**Review by Glomberg — CHANGES_REQUESTED** [review]

File: lib/Cleantalk/Antispam/Cleantalk.php [hunk 3]
```diff
@@ -547,8 +547,8 @@ private function sendRequest($data, $url, $server_timeout = 3)
         }
 
         //common way - left this if we need to specify method uri for 2.0
-        $url .= !empty($this->method_uri) && is_string($this->method_uri)
-            ? '/' . $this->method_uri
+        $url = !empty($this->method_uri) && is_string($this->method_uri)
+            ? $url . '/' . $this->method_uri
```
[the remark this answers is not in the document]
Comment (alexandergull) [reply lib/Cleantalk/Antispam/Cleantalk.php:551]: api2.0 is a part of URL and should not be a GET param, got this verified

File: lib/Cleantalk/Antispam/Cleantalk.php [hunk 4]
```diff
@@ -547,8 +547,8 @@ private function sendRequest($data, $url, $server_timeout = 3)
         }
 
         //common way - left this if we need to specify method uri for 2.0
-        $url .= !empty($this->method_uri) && is_string($this->method_uri)
-            ? '/' . $this->method_uri
+        $url = !empty($this->method_uri) && is_string($this->method_uri)
```
[the remark this answers is not in the document]
Comment (alexandergull) [reply lib/Cleantalk/Antispam/Cleantalk.php:550]: I can not do it right now without refactoring. The api2.0/api3.0 difference is not implemented in the common lib. This chunk use method_uri param and all the logic that use setPresets() logic. We can't just modify the whole common lib not being sure that all the CMS call implemented setPresets() method and the provided preset exists - look for https://github.com/CleanTalk/_antispam/blob/14b70bfd579e7476e31f0de19d9f187dd73984e7/Cleantalk.php#L466

**Pull request merged**